### PR TITLE
[ADD] Added new module web_calendar_custom

### PR DESCRIPTION
**Pull request by lgharib**
Added new module web_calendar_custom

### CODE REVIEW

File: web_calendar_custom/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# © 2018 Savoir-faire Linux
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/LGPL).
+
+{
+    'name': 'Calendar Custom Settings',
+    'version': '11.0.1.0.0',
+    'author': 'Savoir-faire Linux',
+    'maintainer': 'Savoir-faire Linux',
+    'website': 'https://www.savoirfairelinux.com',
```
Comment (nikul-serpentcs) [inline web_calendar_custom/__manifest__.py]: URL should be this `https://github.com/OCA/web`

File: web_calendar_custom/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# © 2018 Savoir-faire Linux
```
Comment (tarteo) [inline web_calendar_custom/__init__.py]: Please change the copyright character to "Copyright"

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    "use strict";
+
+    var rpc = require('web.rpc');
+    var AbstractRenderer = require('web.AbstractRenderer');
+    var core = require('web.core');
+
+    var _t = core._t;
+
+    /* eslint no-extend-native: "error" */
+    String.prototype.toHHMMSS = function () {
```
Comment (tarteo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js]: Why not just add to the class?

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            hours = '0' + hours;
+        }
+        if (minutes < 10) {
+            minutes = '0' + minutes;
+        }
+        return hours + ':' + minutes + ':' + '00';
+    }
+
+    /* Get configuration values from res.config.settings */
+    rpc.query({
```
Comment (tarteo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js]: Put this in `willStart` of `AbstractRenderer`

File: web_calendar_custom/README.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+* Change interval duration time in minutes between 15 min, 30 min or 1 hour (Row duration in minutes)
+* Change starting day hour (Start Time)
+* Change ending day hour (End Time)
+* Change initial scrolling day hour (Start Work Time)
+* Enable\Disable Weekend days (Weekend days are visible)
+* Enable\Disable overlapping events (Allow overlapping events)
+* Choose first day of the week Sunday, Monday or Tuesday (First day of week)
+
+=========================================
+Français : Paramètrage du calendrier Odoo
```
Comment (tarteo) [inline web_calendar_custom/README.rst]: Remove the translation

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                    },
+                    eventAfterAllRender: function () {
+                        $(window).trigger('resize');
+                    },
+                    viewRender: function (view) {
+
+                        /* compute mode from view.name which is either
+                         'month', 'agendaWeek' or 'agendaDay' */
+                        var mode = view
+                        .name === 'month' ? 'month' : (
```
Comment (tarteo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js]: Please fix the indentation here.

File: web_calendar_custom/README.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (tarteo) [inline web_calendar_custom/README.rst:4]: Use the new readme template

**Review by tarteo — CHANGES_REQUESTED** [review]
Are the options also accessible through the attributes in the view definition?
Can you fix the conflicts and take a look at travis

File: web_calendar_custom/static/src/js/web_calendar_custom.js (hunk 4 above)
Comment (lgharib) [reply web_calendar_custom/static/src/js/web_calendar_custom.js]: Hi @tarteo I tried to put the rpc call in willStart and put the rest in Start. Also and changed AbstractRenderer.include to AbstractRenderer.extend but the code won't execute. How can I contact you for some advise?
Regards.

Comment (lgharib) [reply web_calendar_custom/static/src/js/web_calendar_custom.js]: I finally added all the code in willStart and Start function by using include AbstractRenderer to extend the functions used for applying the new settings to the calendar. And it worked. However I still have to use a global variable to pass rpc call results, from willStart function to Start function. Regards, Larbi

**Review by lgharib — COMMENTED** [review]
I updated the code with recommendations from @tarteo and @nikul-serpentcs. I fixed indentation issues used the general template for the readme and removed the french translation. I also refactored the code so that the RPC call is made in the willStart function and the calendar process is done in the Start function.

File: web_calendar_custom/README.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+Authors
+~~~~~~~
+
+* Savoir-faire Linux
+
+
+Contributors
+~~~~~~~~~~~~
+
+Contributors
```
Comment (eilst) [inline web_calendar_custom/README.rst]:
```suggestion
```

File: web_calendar_custom/README.rst [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+------------
```
Comment (eilst) [inline web_calendar_custom/README.rst]:
```suggestion
```

File: web_calendar_custom/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Copyright 2019 Savoir-faire Linux
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/LGPL).
+
+{
+    'name': 'Calendar Custom Settings',
+    'version': '11.0.1.0.0',
+    'author': 'Savoir-faire Linux',
```
Comment (lgharib) [inline web_calendar_custom/__manifest__.py]:
```suggestion
    'author': 'Savoir-faire Linux, '
                 'Odoo Community Association (OCA)',
```

**Review by lgharib — COMMENTED** [review]
Changing the authors in __manifest__.py

File: web_calendar_custom/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
```
Comment (eilst) [inline web_calendar_custom/__manifest__.py]:
```suggestion
    'author': 'Savoir-faire Linux, '
```

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 12]
```diff
@@ -0,0 +1,143 @@
+/*
+    © 2018 Savoir-faire Linux <https://savoirfairelinux.com>
```
Comment (Tardo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js]:
```suggestion
    Copyright 2018 Savoir-faire Linux <https://savoirfairelinux.com>
```

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+    /* Override fullCalendar configuration */
+    AbstractRenderer.include({
+
+        willStart: function () {
+
+            /* Get configuration values from res.config.settings */
+            return rpc.query({
+                model: 'res.config.settings',
+                method: 'get_values',
```
Comment (Tardo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:25]: Use this method can have an impact in the performance... please, refactor the python part to separate the logic that prepare the values.

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+odoo.define('web_calendar_custom.renderer_setting', function (require) {
+    "use strict";
+
+    var rpc = require('web.rpc');
+    var AbstractRenderer = require('web.AbstractRenderer');
+    var core = require('web.core');
+
+    var _t = core._t;
+    var config_values = null;
```
Comment (Tardo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:15]: Please, don't declare variables out of the class scope if is not really necessary. I don't know if you want 'emulate' an static member.

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        },
+
+        start: function () {
+            this._super.apply(this, arguments);
+
+            var self = this;
+
+            /* eslint no-extend-native:
+            ["error", { "exceptions": ["Object"] }] */
+            String.prototype.toHHMMSS = function () {
```
Comment (Tardo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:38]: Perhaps can use 'web.field_utils' (format formatFloatTime) instead

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+                if (minutes < 10) {
+                    minutes = '0' + minutes;
+                }
+                return hours + ':' + minutes + ':' + '00';
+            };
+
+            this.$calendar = this.$(".o_calendar_widget");
+
+            /* Documentation here : http://arshaw.com/fullcalendar/docs */
+            if (typeof this.$calendar.fullCalendar !== 'undefined' &&
```
Comment (Tardo) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:55]: Use ```_.isUndefined(this.$calendar.fullCalendar)``` & ```_.isNull(this.$calendar.fullCalendar)```

**Review by Tardo — CHANGES_REQUESTED** [review]

File: web_calendar_custom/data/ir_config_parameter.xml [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+        <field name="value">7.0</field>
+    </record>
+    <record id="calendar_row_duration" model="ir.config_parameter">
+        <field name="key">web_calendar_custom.calendar_row_duration</field>
+        <field name="value">0.25</field>
+    </record>
+    <record id="calendar_first_day_of_week" model="ir.config_parameter">
+        <field name="key">web_calendar_custom.calendar_first_day_of_week</field>
+        <field name="value">1</field>
+    </record>
```
Comment (daramousk) [inline web_calendar_custom/data/ir_config_parameter.xml:22]: Perhaps you need `is_weekend_active` in here as well?

File: web_calendar_custom/models/res_config_settings.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+         "(calendar_start_work_time < calendar_end_time) AND"
+         "( calendar_start_time < 24) AND (calendar_end_time >= 0))",
+         _('Start work time must be set between start and end time')),
+    ]
+
+    @api.model
+    def get_values(self):
+        res = super(ResConfigSettings, self).get_values()
+        res.update(
+            calendar_start_time=self.env[
```
Comment (daramousk) [inline web_calendar_custom/models/res_config_settings.py:41]: We can try avoid repeating ourselves by `model_ir_config_parameter = self.env['ir.config_parameter'].sudo()` and then use that for all our calls.

File: web_calendar_custom/models/res_config_settings.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+            calendar_row_duration=self.env[
+                'ir.config_parameter'].sudo().get_param(
+                'web_calendar_custom.calendar_row_duration'),
+        )
+        return res
+
+    @api.multi
+    def set_values(self):
+        super(ResConfigSettings, self).set_values()
+        self.env['ir.config_parameter'].sudo().set_param(
```
Comment (daramousk) [inline web_calendar_custom/models/res_config_settings.py:68]: Same here as above

File: web_calendar_custom/views/res_config_views.xml [hunk 20]
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="res_config_settings_view_form" model="ir.ui.view">
+            <field name="name">res.config.settings.view.form.inherit.web.calendar</field>
+            <field name="model">res.config.settings</field>
+            <field name="priority" eval="86"/>
```
Comment (daramousk) [inline web_calendar_custom/views/res_config_views.xml:7]: What is this priority for?

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+
+    var rpc = require('web.rpc');
+    var AbstractRenderer = require('web.AbstractRenderer');
+    var core = require('web.core');
+
+    var _t = core._t;
+    var config_values = null;
+
+    /* Override fullCalendar configuration */
+    AbstractRenderer.include({
```
Comment (daramousk) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:18]: Why inherit from `AbstractRenderer` when you can inherit from `CalendarRenderer`?

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+        willStart: function () {
+
+            /* Get configuration values from res.config.settings */
+            return rpc.query({
```
Comment (daramousk) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:23]: Use `get_param`, `set_param` to get/set params.

File: web_calendar_custom/static/src/js/web_calendar_custom.js [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+                config_values = config_data;
+            });
+        },
+
+        start: function () {
+            this._super.apply(this, arguments);
+
+            var self = this;
+
+            /* eslint no-extend-native:
```
Comment (daramousk) [inline web_calendar_custom/static/src/js/web_calendar_custom.js:36]: `eslint` is right, do not do this, Tardo correctly points out the solution below. Not only you should avoid extending `String` but this function could be a little bit off.

**Review by daramousk — CHANGES_REQUESTED** [review]
Hello there, and thank you for this piece of work, I have left some comments and a conclusion below:

I believe that is more complicated than needed:

Why don't you extend `CalendarRenderer`, `_initCalendar`, `initCalendarMini`, `_render` to do what you need? The code that you will need to write will be significantly less.

**Pull request closed without merging**